### ISSUE TITLE
How to compute loss from this kind of `input_ids` and `labels`?

### ISSUE THREAD

**Issue by fahadh4ilyas** [linfeng93/BiTA#2]
I don't understand the format of `input_ids` and `labels` after preprocessing dataset. Why is the `mask_id` using number bigger than embedding size and vocab size which is 32000? Will it be problem when entering the model which has `Embedding` layer size exactly 32000. But, here is what I got from `input_ids`

```python
[1, 518, 25580, 29962, 3532, 14816, 29903, 6778, 13, 3492, 526, 263, 8444, 29892, 3390, 1319, 322, 15993, 20255, 29889, 29849, 1234, 408, 1371, 3730, 408, 1950, 29892, 1550, 1641, 9109, 29889, 29871, 3575, 6089, 881, 451, 3160, 738, 10311, 1319, 29892, 443, 621, 936, 29892, 11021, 391, 29892, 7916, 391, 29892, 304, 27375, 29892, 18215, 29892, 470, 27302, 2793, 29889, 3529, 9801, 393, 596, 20890, 526, 5374, 635, 443, 5365, 1463, 322, 6374, 297, 5469, 29889, 13, 13, 3644, 263, 1139, 947, 451, 1207, 738, 4060, 29892, 470, 338, 451, 2114, 1474, 16165, 261, 296, 29892, 5649, 2020, 2012, 310, 22862, 1554, 451, 1959, 29889, 960, 366, 1016, 29915, 29873, 1073, 278, 1234, 304, 263, 1139, 29892, 3113, 1016, 29915, 29873, 6232, 2089, 2472, 29889, 13, 29966, 829, 14816, 29903, 6778, 13, 13, 4035, 29873, 1461, 1023, 3694, 322, 736, 278, 1121, 29889, 13, 29929, 29955, 29892, 29871, 29946, 29906, 518, 29914, 25580, 29962, 29871, 4587, 3236, 29892, 306, 29915, 29881, 367, 9796, 304, 1371, 29991, 1763, 23197, 29871, 29929, 29955, 515, 29871, 29946, 29906, 29892, 591, 508, 3763, 23197, 278, 1023, 3694, 29901, 13, 13, 29946, 29906, 448, 29871, 29929, 29955, 353, 448, 29945, 29945, 13, 13, 29902, 4966, 445, 6911, 29991, 2803, 592, 1073, 565, 366, 505, 738, 32000, 32001, 32002, 32000, 32001, 32002, 32000, 32001, 32002, 32000, 32001, 32002, 32000, 32001, 32002, 32000, 32001, 32002, 32000, 32001, 32002, 32000, 32001, 32002, 32000, 32001, 32002, 32000, 32001, 32002, 32000, 32001, 32002, 32000, 32001, 32002, 32000, 32001, 32002, 32000, 32001, 32002, 32000, 32001, 32002, 32000, 32001, 32002, 32000, 32001, 32002, 32000, 32001, 32002, 32000, 32001, 32002, 32000, 32001, 32002, 32000, 32001, 32002, 32000, 32001, 32002, 32000, 32001, 32002, 32000, 32001, 32002, 32000, 32001, 32002, 32000, 32001, 32002, 32000, 32001, 32002, 32000, 32001, 32002, 32000, 32001, 32002, 32000, 32001, 32002, 32000, 32001, 32002, 32000, 32001, 32002, 32000, 32001, 32002, 32000, 32001, 32002, 32000, 32001, 32002, 32000, 32001, 32002, 32000, 32001, 32002, 32000, 32001, 32002, 32000, 32001, 32002, 32000, 32001, 32002, 32000, 32001, 32002, 32000, 32001, 32002, 32000, 32001, 32002, 32000, 32001, 32002, 32000, 32001, 32002, 32000, 32001, 32002, 32000, 32001, 32002, 32000, 32001, 32002, 32000, 32001, 32002, 32000, 32001, 32002, 32000, 32001, 32002, 32000, 32001, 32002, 32000, 32001, 32002, 32000, 32001, 32002, 32000, 32001, 32002, 32000, 32001, 32002]
```

Clearly `input_ids` number is somehow bigger that 31999. And here is the `labels`

```python
[-100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, 3236, 29892, 306, 29892, 306, 29915, 306, 29915, 29881, 29915, 29881, 367, 29881, 367, 9796, 367, 9796, 304, 9796, 304, 1371, 304, 1371, 29991, 1371, 29991, 1763, 29991, 1763, 23197, 1763, 23197, 29871, 23197, 29871, 29929, 29871, 29929, 29955, 29929, 29955, 515, 29955, 515, 29871, 515, 29871, 29946, 29871, 29946, 29906, 29946, 29906, 29892, 29906, 29892, 591, 29892, 591, 508, 591, 508, 3763, 508, 3763, 23197, 3763, 23197, 278, 23197, 278, 1023, 278, 1023, 3694, 1023, 3694, 29901, 3694, 29901, 13, 29901, 13, 13, 13, 13, 29946, 13, 29946, 29906, 29946, 29906, 448, 29906, 448, 29871, 448, 29871, 29929, 29871, 29929, 29955, 29929, 29955, 353, 29955, 353, 448, 353, 448, 29945, 448, 29945, 29945, 29945, 29945, 13, 29945, 13, 13, 13, 13, 29902, 13, 29902, 4966, 29902, 4966, 445, 4966, 445, 6911, 445, 6911, 29991, 6911, 29991, 2803, 29991, 2803, 592, 2803, 592, 1073, 592, 1073, 565, 1073, 565, 366, 565, 366, 505, 366, 505, 738, 505, 738, 916, 738, 916, 5155, 916, 5155, 29889, 5155, 29889, 2]
```

Could you explain the mechanism to compute loss from this?

**Comment by linfeng93** [comment]
@fahadh4ilyas 

In the ```input_ids```, we use ```mask_id``` larger than the vocabulary size as placeholders. For example, if we use 3 mask tokens, the ```mask_id``` would be 32000, 32001, and 32002 respectively.

In the model init function, we construct ```self.mask_tokens = nn.Embedding(mask_num, config.hidden_size)``` as the ```mask_token``` embeddings. In the forward function, ```self.mask_tokens.weight``` (shape: mask_num, hidden_size) is concatenated with ```self.embed_tokens.weight``` (shape: vocab_size, hidden_size) to obtain new ```embed_tokens```, which are then used to obtain the embeddings of the input query tokens and mask tokens using ```F.embedding(input_ids, embed_tokens)```.

Therefore, you will notice that ```mask_id``` only appears in ```input_ids```, while ```mask_id``` does not appear in the ground truth ```labels```.
